### PR TITLE
switch to xml and json output for PBS and LSF 

### PR DESCRIPTION
**Pull request by fkaercher**
- changed qstat output to XML
- changed bjob output to JSON
- fixed issue #67 and #99 
- unit tests for parsing XML and JSON output
- new integration test

### CODE REVIEW

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy [hunk 1]
```diff
@@ -79,16 +82,16 @@ class LSFJobManager extends AbstractLSFJobManager {
     }
 
     protected Map<BEJobID, JobState> queryJobStates(List<BEJobID> jobIDs) {
-        runBjobs(jobIDs).collectEntries { BEJobID jobID, String[] value ->
-            JobState js = parseJobState(value[getPositionOfJobState()])
+        runBjobs(jobIDs).collectEntries { BEJobID jobID, Object value ->
+            JobState js = parseJobState(value["STAT"] as String)
```
[the remark this answers is not in the document]
Comment (fkaercher) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: Yes, I changed it to a map.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy [hunk 2]
```diff
@@ -198,6 +215,75 @@ class LSFJobManager extends AbstractLSFJobManager {
...
     }
+*/
+
+    /**
+     * Used by @getJobDetails to set JobInfo
+     */
+    private GenericJobInfo queryJobInfo(Object jobResult) {
+
+        GenericJobInfo jobInfo
+        BEJobID jobID = new BEJobID(jobResult["JOBID"] as String)
```
[the remark this answers is not in the document]
Comment (fkaercher) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: Yes, here the same.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 3]
```diff
@@ -245,68 +250,70 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
      */
     private Map<BEJobID, GenericJobInfo> processQstatOutput(List<String> resultLines) {
         Map<BEJobID, GenericJobInfo> queriedExtendedStates = [:]
-        Map<String, Map<String, String>> qstatReaderResult = this.readQstatOutput(resultLines.join("\n"))
+        if(resultLines.join("\n").isEmpty()){
```
[the remark this answers is not in the document]
Comment (fkaercher) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: True, that's not necessary.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy [hunk 4]
```diff
@@ -102,27 +105,42 @@ class LSFJobManager extends AbstractLSFJobManager {
...
 
         if (er.successful) {
             if (resultLines.size() >= 1) {
+                String rawJson = resultLines.join("\n")
+                Object parsedJson = new JsonSlurper().parseText(rawJson)
+                List records = (List) parsedJson.getAt("RECORDS")
+                records.each {
+                    BEJobID jobID = new BEJobID(it["JOBID"] as String)
+                    String jobState = it["STAT"]
+                    logger.info(["QStat BEJob line: " + it,
```
[the remark this answers is not in the document]
Comment (fkaercher) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: Hmm not sure. What would you suggest?

[the remark this answers is not in the document]
Comment (vinjana) [reply src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: Don't know. Maybe "debug"? For writing out all output "info" is certainly too verbose. If we had "trace" we might also use this.

Consider: https://stackoverflow.com/questions/2031163/when-to-use-the-different-log-levels

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy [hunk 5]
```diff
@@ -145,56 +140,67 @@ class LSFJobManager extends AbstractLSFJobManager {
...
      */
-    private GenericJobInfo queryJobInfo(BEJobID jobID) {
-        String[] jobDetails = runBjobs([jobID]).get(jobID)
-        GenericJobInfo jobInfo
+    private GenericJobInfo queryJobInfo(Map<String, Object> jobResult) {
 
-        List<String> dependIDs = jobDetails[32].tokenize(/&/).collect { it.find(/\d+/) }
-        jobInfo = new GenericJobInfo(jobDetails[1], new File(jobDetails[31]), jobDetails[0], null, dependIDs)
+        GenericJobInfo jobInfo
+        BEJobID jobID = new BEJobID(jobResult["JOBID"] as String)
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/lsf/LSFJobManager.groovy]: Nice. The code is now much more self-documenting and safe.

File: src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy [hunk 6]
```diff
@@ -234,8 +240,7 @@ class PBSJobManager extends ClusterJobManager<PBSCommand> {
     }
 
     private static LocalDateTime parseTime(String str) {
-        def pbsDatePattern = DateTimeFormatter.ofPattern("EEE MMM ppd HH:mm:ss yyyy").withLocale(Locale.ENGLISH)
-        return LocalDateTime.parse(str, pbsDatePattern)
+        return catchExceptionAndLog {Instant.ofEpochSecond(Long.valueOf(str)).atZone(ZoneId.systemDefault()).toLocalDateTime()}
```
Comment (vinjana) [inline src/main/groovy/de/dkfz/roddy/execution/jobs/cluster/pbs/PBSJobManager.groovy]: Just as remark (no request for change): I know this kind of functions with the "with..." conventions. E.g. 
```
withCaughtAndLoggedExceptions {
   ...
}
```

`catchAndLogExceptions` would be slightly less bumpy :-)

**Pull request merged**